src/common/storage/src/cache.rs

```diff
@@ -29,27 +36,78 @@ use opendal::OpRead;
 use opendal::OpWrite;
 use opendal::Result;
 
-/// TODO: implement more complex cache logic.
-///
-/// For example:
+/// VisitStatistics is used to track visit statistics.
+#[derive(Debug, Clone)]
+pub struct VisitStatistics {
+    cache: Arc<Cache<String, Arc<AtomicUsize>>>,
+}
+
+impl VisitStatistics {
+    /// Create a new visit statistics with given capacity.
+    pub fn new(capacity: u64) -> Self {
+        VisitStatistics {
+            cache: Arc::new(
+                Cache::builder()
+                    .max_capacity(capacity)
+                    // Time to live (TTL): 30 minutes
+                    //
+                    // TODO: make this a user setting.
+                    .time_to_live(Duration::from_secs(30 * 60))
+                    // Time to idle (TTI):  5 minutes
+                    //
+                    // TODO: make this a user setting.
+                    .time_to_idle(Duration::from_secs(5 * 60))
+                    .build(),
+            ),
+        }
+    }
+
+    /// Visit this path.
+    ///
+    /// Returns the previous value
+    fn visit(&self, path: String) -> usize {
+        let v = self.cache.get_with(path, || Arc::new(AtomicUsize::new(0)));
+        v.fetch_add(1, Ordering::Relaxed)
+    }
+}
+
+/// RangeCachePolicy will try to read and store cache based on range.
 ///
-/// - Implement a top n heap, and only cache files exist in heap.
-/// - Only cache data file, and ignore snapshot files.
-#[derive(Debug, Default)]
-pub struct FuseCachePolicy {}
-
-impl FuseCachePolicy {
-    pub fn new() -> Self {
-        FuseCachePolicy::default()
+/// We will count recent `records`, if they have been visited at least `threshold`
+/// times. We will cache it in the cache layer.
+#[derive(Debug, Clone)]
+pub struct RangeCachePolicy {
+    visit: VisitStatistics,
+    threshold: usize,
+
+    enable_async: bool,
+    concurrency: Arc<Semaphore>,
+}
+
+impl RangeCachePolicy {
+    /// Create a new range cache policy.
+    pub fn new(records: u64, threshold: usize) -> Self {
+        RangeCachePolicy {
+            visit: VisitStatistics::new(records),
+            threshold,
+            enable_async: false,
+            concurrency: Arc::new(Semaphore::new(16)),
+        }
+    }
+
+    /// Enable async caching.
+    pub fn enable_async(mut self) -> Self {
+        self.enable_async = true;
+        self
     }
 
     fn cache_path(&self, path: &str, args: &OpRead) -> String {
-        format!("{path}.cache-{}", args.range().to_header())
+        format!("{path}.cache-{}", args.range())
     }
 }
 
 #[async_trait]
-impl CachePolicy for FuseCachePolicy {
+impl CachePolicy for RangeCachePolicy {
```

Comment:
Move this to a single file like `range_cache.rs`?  We can add more about how the range cache policies work to this file.

Reply:
Oh, the file name `cache.rs` is ok.
Add more about how the cache work, like when the lay1 is full and evicted to the lay2.